src/gateway/router/routes/registerContractRoute.ts

```diff
@@ -0,0 +1,174 @@
+import Router from '@koa/router';
+import { evalType } from '../../tasks/contractsMetadata';
+import { BUNDLR_NODE2_URL, BUNDLR_GATEWAY_URL } from '../../../constants';
+import { sleep } from 'warp-contracts';
+import { getCachedNetworkData } from '../../tasks/networkInfoCache';
+import { sendNotificationToCache } from '../../publisher';
+import { evalManifest } from './deployContractRoute';
+import { Tag } from 'arweave/node/lib/transaction';
+import { stringToB64Url } from 'arweave/node/lib/utils';
+import { fetch } from 'undici';
+import { backOff } from 'exponential-backoff';
+import { getTestnetTag } from './deployBundledRoute';
+
+const BUNDLR_QUERY = `query Transactions($ids: [String!]) {
+    transactions(ids: $ids) {
+      edges {
+        node {
+          address
+        }
+      }
+    }
+  }`;
+
+export async function registerContractRoute(ctx: Router.RouterContext) {
+  const { logger, gatewayDb } = ctx;
+
+  let initStateRaw, contractTx;
+  let txId: any;
+
+  try {
+    txId = ctx.request.body.id;
+
+    const txMetadata = ((await getBundlrGqlMetadata(txId)) as any).transactions.edges[0].node;
+
+    const { contractTagsIncluded, tags } = await verifyContractTags(txId);
+    if (!contractTagsIncluded) {
+      ctx.throw(400, 'Bundlr transaction is not valid contract transaction.');
+    }
+
+    logger.debug('Bundlr transaction marked as valid contract transaction.');
+
+    let encodedTags: Tag[] = [];
+
+    for (const tag of tags) {
+      try {
+        encodedTags.push(new Tag(stringToB64Url(tag.name), stringToB64Url(tag.value)));
+      } catch (e: any) {
+        throw new Error(`Unable to encode tag ${tag.name}: ${e.status}`);
+      }
+    }
+
+    const srcTxId = tags.find((t: any) => t.name == 'Contract-Src')!.value;
+    initStateRaw = tags.find((t: any) => t.name == 'Init-State')!.value;
+    const initState = JSON.parse(initStateRaw);
+    const type = evalType(initState);
+    const ownerAddress = txMetadata.address;
+    const contentType = tags.find((t: any) => t.name == 'Content-Type')!.value;
+    const testnet = getTestnetTag(tags);
+    const manifest = evalManifest(tags);
+
+    contractTx = {
+      id: txId,
+      owner: ownerAddress,
+      data: null,
+      signature: '',
```

Comment:
waiting for bundlr response.. currently we dont have an option to retrieve the data item signature (unless we request arweave gateway)

Reply:
done